### PR TITLE
Add helm-template function

### LINKED ISSUE #24: Add a Helm expand source
@pwittrock FYI

### PR DESCRIPTION
**Pull request by prachirp**
- Add helm-template kpt function
- Add e2e test script for this repo at tests/e2e.sh
- Add publish-functions script to publish this repo's functions to catalog
- Copy example-configs from sdk
- Check in two helm charts under sh/test-data/helm for testing. Keep these separate from example-configs since they are not hydrated yaml.

Fixes #24

### CODE REVIEW

File: sh/src/helm-template [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+create_tmp() {
+  tmp=$(mktemp -d "/tmp/helm-template.XXXXXXXX")
+  # If there is input through stdin then write it to tmp
+  if [[ ! -t 0 ]]; then
+    cat "-" | kpt fn sink "$tmp"
+  fi
+}
+
+delete_tmp() {
```
Comment (frankfarzan) [inline sh/src/helm-template]: why do need to delete this given it only exists in the ephemeral container?

File: sh/src/helm-template [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+run_main() {
+  read_arguments "$@"
+  create_tmp
+  expand_helm
+  delete_tmp
+}
+
+if [[ "${BASH_SOURCE[0]}" == "${0}" ]]
```
Comment (frankfarzan) [inline sh/src/helm-template]: why do you need this conditional?

File: sh/src/helm-template [hunk 3]
```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env bash
```
Comment (frankfarzan) [inline sh/src/helm-template]: I suggest organzing this repo by function instead:

helm/
--  Dockerfile
--  helmp-template.sh
-- test

File: sh/build/helm-template.Dockerfile [hunk 4]
```diff
@@ -0,0 +1,13 @@
+FROM alpine/helm:3.1.1
+
+RUN apk add bash curl git
+RUN apk update
+
+RUN curl -fsSL -o /usr/local/bin/kpt https://storage.googleapis.com/kpt-dev/latest/linux_amd64/kpt
+RUN chmod +x /usr/local/bin/kpt
+ENV PATH /usr/local/bin:$PATH
```
Comment (frankfarzan) [inline sh/build/helm-template.Dockerfile]: why is this needed? What's the default value of PATH

File: sh/test-data/helm/mongodb/templates/NOTES.txt [hunk 5]
```diff
@@ -0,0 +1,96 @@
+This Helm chart is deprecated
```
Comment (frankfarzan) [inline sh/test-data/helm/mongodb/templates/NOTES.txt]: ?

File: sh/test-data/helm/mongodb/Chart.yaml [hunk 6]
```diff
@@ -0,0 +1,20 @@
+apiVersion: v1
```
Comment (frankfarzan) [inline sh/test-data/helm/mongodb/Chart.yaml]: This needs to have a licnece? Also if you can avoid checking it in the repo, it's preferred. You can curl it for testing,

File: tests/e2e.sh [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+function assert_dir_exists() {
+  [[ -d $1 ]] || fail "Dir not exist: $1"
+}
+
+############################
+# Docker Tests
+############################
+
+[[ -z ${NODOCKER} ]] || {
```
Comment (frankfarzan) [inline tests/e2e.sh]: Remove this and the variable as this is only does docker testing

File: tests/e2e.sh [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+testcase "docker_helm_template_usage_too_few_args"
+docker run -v "$(pwd)/helm":/source gcr.io/kpt-functions/helm-template:"${TAG}" name=too-few-args 2>err.txt || true
+assert_contains_string err.txt "${HELM_USAGE_SNIPPET}"
+
+testcase "docker_helm_template_usage_too_many_args"
+docker run -v "$(pwd)/helm":/source gcr.io/kpt-functions/helm-template:"${TAG}" chart_path=/source/redis name=too-many-args extra=args_provided 2>err.txt || true
+assert_contains_string err.txt "${HELM_USAGE_SNIPPET}"
+
+testcase "docker_helm_template_correct_args"
+docker run -v "$(pwd)/helm":/source gcr.io/kpt-functions/helm-template:"${TAG}" name=correct-args chart_path=/source/redis >out.yaml
```
Comment (frankfarzan) [inline tests/e2e.sh]: Extract pwd/helm into a variable and reuse

File: tests/e2e.sh [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+assert_dir_exists mongodb
+assert_dir_exists redis
+
+############################
+# kpt fn Tests
+############################
+
+testcase "kpt_helm_template_imperative"
+kpt fn source example-configs |
+  docker run -v "$(pwd)/helm":/source gcr.io/kpt-functions/helm-template chart_path=/source/mongodb name=my-mongodb |
```
Comment (frankfarzan) [inline tests/e2e.sh]: What about using --image?

Also, test using ConfigMap functionConfig?

File: sh/src/helm-template [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+  fi
+}
+
+delete_tmp() {
+  rm -rf "$tmp"
+}
+
+expand_helm() {
+  # Overwrite files if they exist
+  if ! output="$(helm template "${name}" "${chartPath}" -f "${valuesPath}" --output-dir "${tmp}" )"; then
```
Comment (frankfarzan) [inline sh/src/helm-template]: Problem with wrapping helm cli like this is that you're preventing user from specify any other flag to template:

https://helm.sh/docs/helm/helm_template/

Instead, I think you want to only handle and redirect --output.

File: example-configs/README.md [hunk 11]
```diff
@@ -0,0 +1 @@
+Example k8s configs.
```
[the remark this answers is not in the document]
Comment (prachirp) [reply example-configs/README.md]: Done

File: sh/src/helm-template (hunk 1 above)
Comment (prachirp) [reply sh/src/helm-template]: Good point, I was just following it as a best practice. Removed the delete_tmp function now.

File: sh/src/helm-template (hunk 2 above)
Comment (prachirp) [reply sh/src/helm-template]: It's a way to tell if the script is being sourced. But this script will always be executed in a docker container, never sourced, so I removed the conditional.

File: sh/src/helm-template (hunk 3 above)
Comment (prachirp) [reply sh/src/helm-template]: Changed, also removed test data based on below comment.

File: sh/build/helm-template.Dockerfile (hunk 4 above)
Comment (prachirp) [reply sh/build/helm-template.Dockerfile]: It's a best practice from Docker docs, "use ENV to update the PATH environment variable for the software your container installs"

File: sh/test-data/helm/mongodb/templates/NOTES.txt (hunk 5 above)
Comment (prachirp) [reply sh/test-data/helm/mongodb/templates/NOTES.txt]: They moved to the bitnami charts repo 2 days ago. I now use the bitnami charts repo.

File: sh/test-data/helm/mongodb/Chart.yaml (hunk 6 above)
Comment (prachirp) [reply sh/test-data/helm/mongodb/Chart.yaml]: The bitnami charts use the Apache 2.0 license, but that's fair that we want to avoid checking it in. I now get it only for testing.

File: tests/e2e.sh (hunk 7 above)
Comment (prachirp) [reply tests/e2e.sh]: Done

File: tests/e2e.sh (hunk 8 above)
Comment (prachirp) [reply tests/e2e.sh]: It doesn't really make sense extracting the $(pwd) since we'd have to declare a new variable for each test case.  I extracted the common parts of the path to a variable.

File: tests/e2e.sh (hunk 9 above)
Comment (prachirp) [reply tests/e2e.sh]: Using --image doesn't work, working with @pwittrock to investigate.

File: sh/src/helm-template (hunk 10 above)
Comment (prachirp) [reply sh/src/helm-template]: I agree we prevent the user from specifying options. I'd rather check in this MVP first and address this in a separate change.

File: sh/build/helm-template.Dockerfile (hunk 4 above)
Comment (frankfarzan) [reply sh/build/helm-template.Dockerfile]: But you're using a well-known bin directory, not a user-defined one.

File: tests/e2e.sh (hunk 9 above)
Comment (frankfarzan) [reply tests/e2e.sh]: This is about supporting running priviledged containers

File: sh/src/helm-template (hunk 10 above)
Comment (frankfarzan) [reply sh/src/helm-template]: That's fine. I would look into parsing $@ and passing the rest to Helm. This script should ideally only know about the output flag. In fact, in the most general form, this script can be reused for wrapping any CLIs like Helm template.

File: sh/helm-template/helm-template [hunk 12]
```diff
@@ -0,0 +1,92 @@
+#!/usr/bin/env bash
```
Comment (frankfarzan) [inline sh/helm-template/helm-template]: 1. This file needs a .sh extension
2. Directory name already qualifies this script, so rename it to something else like `run.sh`.
3. Rename top-level `sh` dir to `functions`

Comment (frankfarzan) [inline sh/helm-template/helm-template]: Remember to add LICENSE headers to all new files

File: sh/helm-template/helm-template.Dockerfile [hunk 13]
```diff
@@ -0,0 +1,13 @@
+FROM alpine/helm:3.1.1
```
Comment (frankfarzan) [inline sh/helm-template/helm-template.Dockerfile]: rename this to just Dockerfile

File: sh/src/helm-template (hunk 10 above)
Comment (frankfarzan) [reply sh/src/helm-template]: I'm fine with following up on a separate PR

File: tests/e2e.sh (hunk 9 above)
Comment (prachirp) [reply tests/e2e.sh]: Added todo

File: sh/helm-template/helm-template (hunk 12 above)
Comment (prachirp) [reply sh/helm-template/helm-template]: Done

File: sh/helm-template/helm-template.Dockerfile (hunk 13 above)
Comment (prachirp) [reply sh/helm-template/helm-template.Dockerfile]: Done

File: sh/helm-template/helm-template (hunk 12 above)
Comment (prachirp) [reply sh/helm-template/helm-template]: Done

File: scripts/publish-functions.sh [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+
+REPO="$(cd "$(dirname "${BASH_SOURCE[0]}")"/.. >/dev/null 2>&1 && pwd)"
+TAG=${TAG:-dev}
+FUNCS=(helm-template)
+
+for FN_NAME in "${FUNCS[@]}"; do
+  DOCKER_IMAGE="gcr.io/kpt-functions/${FN_NAME}:${TAG}"
+  DOCKER_FILE="${REPO}/functions/${FN_NAME}/Dockerfile"
+  docker build -t "$DOCKER_IMAGE" -f "$DOCKER_FILE" functions
+	docker push "$DOCKER_IMAGE"
```
Comment (frankfarzan) [inline scripts/publish-functions.sh]: indentation issue

File: tests/e2e.sh [hunk 15]
```diff
@@ -67,6 +81,7 @@ assert_dir_exists redis
 # kpt fn Tests
 ############################
 
+# TODO: Add kpt_helm_template_short test after kpt fn run --image fix <https://github.com/GoogleContainerTools/kpt/issues/359>
```
Comment (frankfarzan) [inline tests/e2e.sh]: Also a test for declarative run

File: scripts/publish-functions.sh (hunk 14 above)
Comment (prachirp) [reply scripts/publish-functions.sh]: Done

File: tests/e2e.sh (hunk 15 above)
Comment (prachirp) [reply tests/e2e.sh]: Done
